---
title: "chromDensity.gnomAD_pSTR"
---

```{r knitr, include = FALSE}
DOCNAME = "chromDensity.gnomAD_pSTR"
dir.create(here::here("output", DOCNAME), showWarnings = FALSE)

NOW <- Sys.time()
# Time chunks during knitting
knitr::knit_hooks$set(timeit = function(before) {
    if (before) {
        print(paste("Start:", Sys.time()))
        NOW <<- Sys.time()
    } else {
        print(paste("Stop:", Sys.time()))
        print(Sys.time() - NOW)
    }
})

knitr::opts_chunk$set(
    autodep        = TRUE,
    cache.lazy     = FALSE, 
    cache.comments = FALSE,
    echo           = TRUE,
    error          = FALSE,
    fig.align      = "center",
    fig.width      = 10,
    fig.height     = 8,
    message        = FALSE,
    warning        = FALSE,
    timeit         = TRUE
)
```

## Setup

Load packages

```{r libaries, cache = FALSE, message=FALSE, warning=FALSE}
# Tidyverse
library(tidyverse)

# plot
library(scales)
library(patchwork)
library(cowplot)
library(ggpubr)
library(ggcorrplot)
library(ggrepel)
library(ggrastr)
library(EnvStats)
theme_set(theme_cowplot(font_size = 12, # axis.title
                        rel_small = 10/12, # axis.text
                        rel_tiny = 9/12,
                        rel_large = 12/12))

# cmplot
library(CMplot)

# heatmap
library(pheatmap)

# color
library(ggsci)
library(RColorBrewer)

# color
cPalette <- c("#999999", "#79AF97FF", "#EFC000FF", "#7AA6DCFF", "#f0b98d", "#CD534CFF")
names(cPalette) = c('All', 'p2', 'p3', 'p4', 'p5', "p6")
```

Load R source

```{r source, cache = FALSE}
source(here::here("code/color.R"))
```

## Load data

load count

```{r load-data, message=FALSE, warning=FALSE}
dat_dir = '/niuyw-usb-disk/Projects/STR/chrom_density/gnomAD'

dat = read_delim(file.path(dat_dir, 'pSTR.bin_count'), delim = '\t', col_names = F)
colnames(dat) = c('chr', 'start', 'end', 'p2', 'p3', 'p4', 'p5', 'p6')
```

sum

```{r}
sum(dat$p2)
sum(dat$p3)
sum(dat$p4)
sum(dat$p5)
sum(dat$p6)
```


load centromere

```{r load-centromere, message=FALSE, warning=FALSE}
cents = read_delim(file.path(dat_dir, 'centromeres.hg38.bed'), delim = '\t', col_names = F)
cents = cents[,-4]
colnames(cents) = c('chr', 'start', 'end')
```

## Neat data

```{r neat-data}
# modified from: https://github.com/talkowski-lab/gnomad-sv-pipeline/blob/master/gnomad_sv_manuscript_code/plotting_code/chromosome_sv_density_analysis.R
neatPileup <- function(dat=NULL, cents=NULL, window=11, smooth=T){
  #Sum counts
  dat$All <- apply(dat[, -c(1:3)], 1, sum, na.rm=T)
  #Smooth counts per chromosome
  require(zoo, quietly=T)
  if(smooth==T){
    for(k in unique(dat$chr)){
      dat[which(dat$chr==k), -c(1:3)] <- apply(dat[which(dat$chr==k), -c(1:3)], 2, function(vals){
        rollapply(as.numeric(vals), width=window, FUN=mean, partial=T)
      })
    }
  }
  #Assign normalized distance from centromere
  #Range: -1 = at pter, 0 = in centromere, +1 = at qter
  dat$cdist.norm <- as.numeric(unlist(sapply(unique(dat$chr), function(k){
    midpoints <- (dat$start[which(dat$chr==k)]+dat$end[which(dat$chr==k)])/2
    c.mid <- mean(c(min(cents[which(cents$chr==k), 2]), max(cents[which(cents$chr==k), 3])))
    c.dists <- midpoints-c.mid
    plen <- c.mid
    qlen <- max(midpoints)-c.mid
    c.dists.n <- c.dists
    c.dists.n[which(c.dists<0)] <- c.dists.n[which(c.dists<0)]/plen
    c.dists.n[which(c.dists>0)] <- c.dists.n[which(c.dists>0)]/qlen
    c.dists.n[which(c.dists.n< -1)] <- -1
    c.dists.n[which(c.dists.n>1)] <- 1
    return(c.dists.n)
  })))
  return(dat)
}

# Gather meta-chromosome average for a single SVTYPE
metaAverage <- function(dat, SVTYPE="All", n.bins=250){
  dat = as.data.frame(dat)
  p.bins <- seq(-1, 0, by=1/n.bins)
  q.bins <- seq(0, 1, by=1/n.bins)
  col.idx <- which(colnames(dat)==SVTYPE)
  p.means <- sapply(1:(length(p.bins)-1), function(i){
    mean(dat[which(dat$cdist.norm>=p.bins[i] & dat$cdist.norm<p.bins[i+1]), col.idx], na.rm=T)
  })
  q.means <- sapply(1:(length(q.bins)-1), function(i){
    mean(dat[which(dat$cdist.norm>q.bins[i] & dat$cdist.norm<=q.bins[i+1]), col.idx], na.rm=T)
  })
  means <- c(p.means, q.means)
  means.norm <- means/mean(means, na.rm=T)
  out.df <- data.frame("norm.pos"=c(p.bins[-length(p.bins)], q.bins[-1]), 
                       "mean"=means, "mean.norm"=means.norm)
  return(out.df)
}

# Split densities into terminal, interstitial, and pericentromeric bins
calc.meanByContext <- function(dat, meta.svtypes, ter.buf=0.05, cen.buf=0.05){
  #Helper function to calculate mean, 95% CI, and p-value that the true mean isn't 1
  get.ci <- function(vals){
    vals <- vals[which(!is.na(vals))]
    k <- 1.96*(sd(vals, na.rm=T)/sqrt(length(vals)))
    p.less <- t.test(vals, mu=1, alternative="less")$p.value
    p.greater <- t.test(vals, mu=1, alternative="greater")$p.value
    return(c(log2(c(mean(vals)-k, mean(vals), mean(vals)+k)), p.less, p.greater))
  }
  res <- lapply(meta.svtypes, function(svtype){
    #Mean-normalize all values
    vals <- as.numeric(dat[, which(colnames(dat)==svtype)])
    vals <- vals/mean(vals, na.rm=T)
    #Calculate stats
    ter.idx <- which(dat$cdist.norm<=-1+ter.buf | dat$cdist.norm>=1-ter.buf)
    cen.idx <- which(dat$cdist.norm>=-cen.buf & dat$cdist.norm<=cen.buf)
    int.idx <- which(!(1:nrow(dat) %in% c(ter.idx, cen.idx)))
    ter.stats <- get.ci(vals[ter.idx])
    int.stats <- get.ci(vals[int.idx])
    cen.stats <- get.ci(vals[cen.idx])
    return(data.frame("ter"=ter.stats, 
                      "int"=int.stats, 
                      "cen"=cen.stats))
  })
  names(res) <- meta.svtypes
  return(res)
}

# Master function to plot raw coverage from all four samples per contig
generateCovPlotsPerChrom <- function(mat, 
                                     colors, 
                                     contigs.top, 
                                     contigs.middle, 
                                     contigs.bottom, 
                                     labels.on.top=F, 
                                     fill=T, norm=F){
  #Normalize data, if optioned
  if(norm==T){
    mat <- lapply(mat, function(chr.mat){
      chr.mat[, -c(1:3)] <- apply(as.data.frame(chr.mat[, -c(1:3)]), 2, function(vals){
        vals <- vals/mean(vals, na.rm=T)
        return(vals)
      })
      return(chr.mat)
    })
  }
  
  #Set spacer
  spacer <- 30000000
  
  #Determine total length to be plotted
  chr.lengths <- as.numeric(unlist(lapply(mat, function(chr.mat){
    return(max(chr.mat[, 3]))
  })))
  #Add spacer distance between contigs
  chr.lengths[1:22] <- chr.lengths[1:22]+spacer
  
  #Mini helper function to plot a row of summary coverage values
  plotMiniCovSummary <- function(contigs, ymax=NULL){
    #Prep plot area
    if(labels.on.top==T){
      par(mar=c(0.2, 2, 1.5, 1), bty="n", bg="white")
      lpos=3
    }else{
      par(mar=c(1.5, 2, 0.2, 1), bty="n", bg="white")
      lpos=1
    }
    if(is.null(ymax)){
      ymax <- as.integer(ceiling(quantile(unlist(lapply(mat, function(chr.mat){chr.mat[, 4]})), probs=0.995)))
    }
    plot(x=c(-0.01*sum(chr.lengths[contigs]), sum(chr.lengths[contigs])-spacer), 
         y=c(-0.1*ymax, 1.1*ymax), type="n", 
         xaxt="n", yaxt="n", xlab="", ylab="", xaxs="i")
    #Add background shading rectangle
    rect(xleft=par("usr")[1], xright=par("usr")[2], 
         ybottom=par("usr")[3], ytop=par("usr")[4], 
         border="gray99", col="gray99")
    box(which="plot", col="white", lwd=3)
    #Add contig positions & labels
    sapply(1:length(contigs), function(i){
      axis(lpos, at=c(c(0, cumsum(chr.lengths[contigs]))[i], 
                      c(0, cumsum(chr.lengths[contigs])-spacer)[i+1]), 
           labels=NA, col="gray30", tck=0, line=0.1)
      axis(lpos, at=mean(c(c(0, cumsum(chr.lengths[contigs]))[i], 
                           c(0, cumsum(chr.lengths[contigs])-spacer)[i+1])), 
           tick=F, labels=paste("chr", contigs[i], sep=""), line=-0.5)
    })
    #Add y axis & gridlines
    y.at <- axTicks(2)[seq(1, length(axTicks(2)), by=2)]
    # y.at <- c(y.at, max(y.at)+(y.at[2]-y.at[1]))
    axis(2, at=axTicks(2), labels=NA, tck=-0.05, col="gray40")
    axis(2, at=y.at, labels=NA, tck=-0.1)
    axis(2, at=y.at, tick=F, las=2, cex.axis=1, line=-0.4)
    # mtext(2, line=2.25, text="Copy Number")
    abline(h=axTicks(2), col="gray80", lwd=0.7, lty=3)
    # abline(h=y.at, col="gray80")
    abline(h=0)
    #Add coverage values
    sapply(1:length(contigs), function(i){
      sapply(1:(ncol(mat[[1]])-3), function(s){
        # cov.vals <- rollmean(cov[[s]][[contigs[i]]][, 4], k=11, na.pad=T)
        vals <- mat[[contigs[i]]][, s+3]
        plot.vals <- smooth.spline(x=mat[[contigs[i]]][, 2]+c(0, cumsum(chr.lengths[contigs]))[i], 
                                   y=vals, spar=0.32*mean(c(1, rep(chr.lengths[contigs[1]]/chr.lengths[contigs[i]], 4))))
        if(fill==T){
          polygon(x=c(plot.vals$x, rev(plot.vals$x)), 
                  y=c(plot.vals$y, rep(0, times=length(plot.vals$y))), 
                  border=NA, col=adjustcolor(colors[s], alpha=0.7))
        }
        points(plot.vals$x, plot.vals$y, type="l", lwd=1.25, col=colors[s])
      })
    })
    #Add cleanup rectangles
    rect(xleft=par("usr")[1], xright=par("usr")[2], 
         ybottom=c(par("usr")[3], ymax), ytop=c(0, par("usr")[4]), 
         border="white", lty=1, lwd=1, col="white")
    abline(h=par("usr")[3:4], lwd=1, col="white")
    # abline(h=c(0, ymax), col="gray80")
    box(lwd=2, col="white")
    rect(xleft=(cumsum(chr.lengths[contigs])-spacer), 
         xright=cumsum(chr.lengths[contigs]), 
         ybottom=par("usr")[3], ytop=par("usr")[4], 
         border="white", lty=0, lwd=0, col="white")
    rect(xleft=c(par("usr")[1], tail((cumsum(chr.lengths[contigs])-spacer), 1)), 
         xright=c(0, par("usr")[2]), 
         ybottom=par("usr")[3], ytop=par("usr")[4], 
         border="white", lty=0, lwd=0, col="white")
  }
  
  #####Plot stacked panels
  par(mfrow=c(3, 1))
  plotMiniCovSummary(contigs.top)
  plotMiniCovSummary(contigs.middle)
  plotMiniCovSummary(contigs.bottom)
}

### Plot meta-chromosome density for a single SVTYPE
plot.metaDist <- function(meta.dat, color, fill=T, norm=F, xlabel=NULL){
  #Clean meta dat
  meta.dat <- meta.dat[which(!is.na(meta.dat$mean)), ]
  if(norm==T){
    meta.dat$mean <- meta.dat$mean.norm
  }
  plot.vals <- rollapply(data=meta.dat$mean.norm, width=6, mean, partial=T)
  #Set parameters
  ymax <- 1.075*max(max(plot.vals, na.rm=T), 
                    2*mean(plot.vals, na.rm=T))
  lpos <- 1
  #Prep plot area
  par(bty="n", bg="white")
  plot(x=1.015*range(meta.dat$norm.pos), y=c(-0.05*ymax, ymax), type="n", 
       xaxt="n", yaxt="n", xlab="", ylab="", xaxs="i")
  #Add background shading rectangle
  rect(xleft=par("usr")[1], xright=par("usr")[2], 
       ybottom=par("usr")[3], ytop=par("usr")[4], 
       border="gray99", col="gray99")
  box(which="plot", col="white", lwd=3)
  #Add contig positions & labels
  axis(lpos, at=par("usr")[1:2], 
       labels=NA, col="gray30", tck=0, line=0.1)
  if(is.null(xlabel)){
    axis(lpos, at=mean(par("usr")[1:2]), 
         tick=F, labels="Meta-chromosome", line=-0.6)
  }else{
    axis(lpos, at=mean(par("usr")[1:2]), 
         tick=F, labels=xlabel, line=-0.6, cex.axis=0.8)
  }
  #Add y axis & gridlines
  y.at <- axTicks(2)
  y.at <- c(y.at, max(y.at)+(y.at[2]-y.at[1]))
  axis(2, at=y.at, labels=NA, tck=-0.05, col="gray40")
  # axis(2, at=y.at, labels=NA, tck=-0.1)
  axis(2, at=y.at, tick=F, las=2, cex.axis=1, line=-0.4, labels=round(y.at, 2))
  # mtext(2, line=2.25, text="Copy Number")
  abline(h=y.at, col="gray80", lwd=0.7, lty=3)
  # abline(h=y.at, col="gray80")
  # abline(v=0, col="#963231")
  if(norm==T){
    abline(h=1, col="gray50")
  }
  #Add coverage values
  # plot.vals <- smooth.spline(x=meta.dat$norm.pos, y=meta.dat$mean, spar=0.2)
  # plot.vals <- data.frame("x"=meta.dat$norm.pos, 
  # "y"=meta.dat$mean)
  if(fill==T){
    polygon(x=c(meta.dat$norm.pos, rev(meta.dat$norm.pos)), 
            y=c(plot.vals, rep(0, times=length(plot.vals))), 
            border=NA, col=adjustcolor(color, alpha=0.7))
  }
  points(meta.dat$norm.pos, plot.vals, type="l", lwd=1.25, col=color)
  #Cleanup
  rect(xleft=par("usr")[1], xright=par("usr")[2], 
       ybottom=par("usr")[3], ytop=0, 
       bty="n", border=NA, col="white")
  abline(h=0)
}

### Plot point estimates and CIs for ter/int/cen averages by class
plot.metaByContext <- function(dat, meta.svtypes, colors){
  dat = as.data.frame(dat)
  #Get point estimates and CIs
  plot.vals <- calc.meanByContext(dat, meta.svtypes)
  #Prep plot area
  ylims <- c(1.1*min(as.numeric(unlist(lapply(plot.vals, range, na.rm=T)))), 
             1.1*max(as.numeric(unlist(lapply(plot.vals, range, na.rm=T)))))
  par(bty="n", mar=c(0.25, 2.8, 1.25, 0.25))
  plot(x=c(0, length(plot.vals)-0.4), y=ylims, type="n", 
       xaxt="n", yaxt="n", xlab="", ylab="")
  abline(h=0, col="gray50")
  #Add points per svtype
  sapply(1:length(plot.vals), function(i){
    segments(x0=i-c(1, 0.8, 0.6), 
             x1=i-c(1, 0.8, 0.6), 
             y0=as.numeric(plot.vals[[i]][1, ]), 
             y1=as.numeric(plot.vals[[i]][3, ]), 
             col=colors[i], lwd=2)
    points(x=i-c(1, 0.8, 0.6), 
           y=as.numeric(plot.vals[[i]][2, ]), 
           pch=19, col=colors[i], cex=1.25)
    text(x=i-c(1, 0.8, 0.6), 
         y=as.numeric(plot.vals[[i]][2, ]), 
         labels=c("T", "I", "C"), cex=0.6, font=2, col="white")
    #Add category label
    axis(3, at=i-c(1, 0.6), tck=0, labels=NA, line=0.1)
    axis(3, at=i-0.8, tick=F, line=-0.9, cex.axis=0.75, labels=meta.svtypes[i], col.axis=colors[i])
    #Add p-values
    par(xpd=T)
    sapply(1:3, function(k){
      if(plot.vals[[i]][4, k]<0.05/(3*length(plot.vals))){
        text(x=(i-c(1, 0.8, 0.6))[k], 
             y=plot.vals[[i]][2, k], 
             pos=1, labels="*")
      }
      if(plot.vals[[i]][5, k]<0.05/(3*length(plot.vals))){
        text(x=(i-c(1, 0.8, 0.6))[k], 
             y=plot.vals[[i]][2, k]-(0.04*(par("usr")[4]-par("usr")[3])), 
             pos=3, labels="*")
      }
    })
    par(xpd=F)
  })
  
  #Clean up
  axis(2, at=axTicks(2), labels=NA, tck=-0.03)
  sapply(axTicks(2), function(y){
    axis(2, at=y, tick=F, las=2, cex.axis=0.9, line=-0.5, 
         labels=bquote("2"^.(y)))
  })
  mtext(2, line=1.75, text="SV Fold-Enrichment")
}
```

Preprocess data

```{r process, message=FALSE, warning=FALSE}
# Process input data
dat = neatPileup(dat = dat, cents = cents, smooth = T)

# Gather meta-chromosome averages per percentile
alt_types = c('All', 'p2', 'p3', 'p4', 'p5', "p6")
meta.means <- lapply(alt_types, function(svtype){
  m <- metaAverage(dat=dat, SVTYPE=svtype, n.bins=500)
})
names(meta.means) <- alt_types

# write dat to use cdist.norm
write_delim(dat, file = here::here("output", DOCNAME, 'chrom_density.dat.txt'), delim = '\t')
```

Output the binning of chromosomes

把 cdist.norm 切成了 40 份，每份 0.05。然后把这个区间输出，以比较这些区间里 pSTR 的特征

```{r}
dat %>%
  mutate(bin = cut(cdist.norm, breaks = seq(-1, 1, 0.05),
                   include.lowest = T, right = F,
                   labels = as.character(setdiff(seq(-20, 20), c(0)))
                   )
         ) %>%
  #dplyr::select(chr, start, end, bin) %>%
  group_by(chr, bin) %>% 
  summarise(left = min(start), right = max(end)) %>%
  ungroup() %>%
  dplyr::select(chr, left, right, bin) %>%
  write_delim(file = here::here("output", DOCNAME, 'hg38.autosomes.bin_100k.cdist.norm_bin.bed'), delim = '\t')
```

## Plot

### Generate "meta-chromosome" plots per period

Large plot: all MEIs

```{r p-3}
par(mar=c(1.5, 2, 0.2, 1))
plot.metaDist(meta.dat=meta.means[[which(names(meta.means)=="All")]], 
              color=cPalette[1], norm=T, 
              xlabel="Meta-chromosome (Mean of Autosomes)")
```

Each period

```{r p-4, fig.width=7, fig.height=7}
par(mfrow=c(3, length(alt_types[-1])/2), 
    mar=c(1.5, 2, 1.5, 1))
lapply(alt_types[-1], function(svtype){
  svt.col <- cPalette[which(names(cPalette) == svtype)]
  plot.metaDist(meta.dat=meta.means[[which(names(meta.means)==svtype)]], 
                color=svt.col, norm=T)
})
```

### Plot meta means by ter/int/cen

```{r p-5}
### Plot meta means by ter/int/cen
plot.metaByContext(dat, alt_types, 
          colors=cPalette)
```

使用 ggplot 绘图

```{r fig.height=2, fig.width=4}
plot.vals = calc.meanByContext(as.data.frame(dat), alt_types)

df_list = lapply(1:length(plot.vals), function(i) {
  x = as.data.frame(t(plot.vals[[i]]))
  colnames(x) = c('lower', 'value', 'upper', 'p_lower', 'p_upper')
  x = x %>%
    rownames_to_column('loci') %>%
    mutate(alt_type = names(plot.vals)[i])
  x
})

df = do.call(rbind, df_list) %>%
  mutate(loci = factor(loci, levels = c('ter', 'int', 'cen'))) %>%
  mutate(alt_type = factor(alt_type, levels = c('All', 'p2', 'p3', 'p4', 'p5', 'p6')))

# to exponent
df$value = 2^(df$value)
df$lower = 2^(df$lower)
df$upper = 2^(df$upper)

p = df %>%
  ggplot(aes(x = alt_type, y = value, color = loci)) +
  geom_pointrange(aes(ymin = lower, ymax = upper), position = position_dodge(.7), size = .8, fatten = 1) +
  scale_color_jco() +
  scale_y_log10(breaks = trans_breaks("log2", function(x) 2^x),
                labels = trans_format("log2", math_format(2^.x))) +
  geom_hline(yintercept = 1, linetype = 'dashed') +
  labs(y = 'SV Fold−Enrichment') +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1))
p
```

### Genome-wide panel of idiograms

All

```{r p-1}
generateCovPlotsPerChrom(mat=lapply(1:22, function(k){
  return(as.data.frame(dat[which(dat$chr == paste0('chr', k)),
                           c(1:3, which(colnames(dat)=="All"))]))}), 
                         colors="gray10", 
                         contigs.top=1:5, 
                         contigs.middle=6:12,
                         contigs.bottom=13:22, 
                         fill=T)
```

Period 6

```{r p-p6}
generateCovPlotsPerChrom(mat=lapply(1:22, function(k){
  return(as.data.frame(dat[which(dat$chr == paste0('chr', k)),
                           c(1:3, which(colnames(dat)=="p6"))]))}), 
                         colors="gray10", 
                         contigs.top=1:5, 
                         contigs.middle=6:12,
                         contigs.bottom=13:22, 
                         fill=T)
```

## Save

"meta-chromosome" plots per period

```{r}
pdf(here::here("output", DOCNAME, "Fig.pSTR_density_metaIdiograms.perPeriod.pdf"),
    height = 6.5, width = 6.8)

par(mfrow=c(3, length(alt_types[-1])/2), 
    mar=c(1.5, 2, 1.5, 1))
lapply(alt_types[-1], function(svtype){
  svt.col <- cPalette[which(names(cPalette) == svtype)]
  plot.metaDist(meta.dat=meta.means[[which(names(meta.means)==svtype)]], 
                color=svt.col, norm=T)
})

dev.off()

```

meta means by ter/int/cen

```{r}
pdf(here::here("output", DOCNAME, "Fig.pSTR_density_meta.byChromContext.dotplots.pdf"),
    height = 2, width = 3.2)

plot.metaByContext(dat, alt_types, 
          colors=cPalette)

dev.off()
```

Genome-wide panel of idiograms for Period 6

```{r}
#pdf(here::here("output", DOCNAME, "Fig.pSTR_density_idiograms.Period6.pdf"),
#    height = 5, width = 7)

png(here::here("output", DOCNAME, "Fig.pSTR_density_idiograms.Period6.png"), 
    height=800, width=(2/3)*2600, res=300, bg="white")
generateCovPlotsPerChrom(mat=lapply(1:22, function(k){
  return(as.data.frame(dat[which(dat$chr == paste0('chr', k)),
                           c(1:3, which(colnames(dat)=="p6"))]))}), 
                         colors="gray10", 
                         contigs.top=1:5, 
                         contigs.middle=6:12,
                         contigs.bottom=13:22, 
                         fill=T)

dev.off()
```

## Session info

