#!/bin/bash
#PBS -V
#PBS -j eo
#PBS -N chr22.FilteredRaw
#PBS -q Fat
#PBS -l nodes=1:ppn=4
#PBS -d /home2/niuyw/project/STR/02_FilteredRaw220120/logs

# WORK DIR
WORKDIR=/home2/niuyw/project/STR/02_FilteredRaw220120

# chrom
echo $chrom

# PPN
PPN=4

# genome and annotations
ANNODIR=/home2/niuyw/RefData
GRCh38_REFERENCE=$ANNODIR/Homo_sapiens/GRCh38_no_alt/genome.fa
GRCh38_GTF=$ANNODIR/Homo_sapiens/GENCODE_v39/gencode.v39.annotation.gtf
GRCh38_bowtie2=$ANNODIR/Homo_sapiens/GRCh38_no_alt/Bowtie2Index/GRCh38_no_alt
GRCh38_STAR=$ANNODIR/Homo_sapiens/GRCh38_no_alt/STARgenomes

# tools dir
TOOLDIR=/home2/niuyw/software
path2java=$TOOLDIR/jdk1.8.0_311/bin/java
path2bedtools=$TOOLDIR/bedtools2/bin/bedtools
path2samtools=$TOOLDIR/samtools.1.9/bin/samtools
path2bcftools=$TOOLDIR/bcftools-1.14/bcftools
path2picard=$TOOLDIR/picard-2.9.2/picard.jar

echo Start time is `date +%Y/%m/%d--%H:%M`

input=/Parastor300s_G30S/shiyr/gangSTR/process-all/3-dumpSTR/${chrom}-dump.vcf.gz

# filter
$path2bcftools view --threads $PPN -S /home2/niuyw/project/STR/samples/sam6487.lst -c 1 -f PASS -Oz -o $WORKDIR/tmp.${chrom}.vcf.gz $input

# add AC, AN, AF, MAF
export BCFTOOLS_PLUGINS=$TOOLDIR/bcftools-1.14/plugins
$path2bcftools +fill-tags --threads $PPN -Oz -o $WORKDIR/${chrom}.vcf.gz $WORKDIR/tmp.${chrom}.vcf.gz -- -t AC,AN,AF,MAF

# index
$TOOLDIR/htslib.1.14/bin/tabix -p vcf $WORKDIR/${chrom}.vcf.gz

# clean
#rm $WORKDIR/tmp.${chrom}.vcf.gz

echo Finish time is `date +%Y/%m/%d--%H:%M`

